### PR TITLE
Automatically edit releases in Confluence

### PR DESCRIPTION
**Pull request by zach-bob**
This PR implements a feature that we have been striving to achieve for awhile: Automating editing in Confluence. To achieve this, lots of refactoring is necessary.

Per [Jira Ticket 52](https://mojotech.atlassian.net/jira/software/projects/MRNN/boards/22?selectedIssue=MRNN-52) and [Jira Ticket 68](https://mojotech.atlassian.net/jira/software/projects/MRNN/boards/22?selectedIssue=MRNN-68), this PR implements functionality for automating the editing releases on Confluence that have already been published on Github if they are edited in Github.

This PR implements persistence tracking of published releases on a persistence provider by using a join table between webhook_events and persistence_providers. The persisted slug is also included in that join table. This method was chosen in favor of directly querying Confluence for a release note at the time an edit event is received.

### CODE REVIEW

File: lib/release_notes_bot/persists.ex [hunk 1]
```diff
@@ -97,6 +143,23 @@ defmodule ReleaseNotesBot.Persists do
     end
   end
 
+  def select_provider_to_update(persistence_provider_id, sanitizer_changes) do
+    case persistence_provider_id do
+      # Confluence
+      1 ->
```
Comment (chewnoill) [inline lib/release_notes_bot/persists.ex:156]: hmmm, it might make sense to add a provider type to this table. Maybe an enum? Probably not important until we have more then one provider type to update.

Comment (zach-bob) [reply lib/release_notes_bot/persists.ex:156]: Would an Enum for this be acceptable as a variable scoped to this file `persists.ex`.
Something like: `@provider_type %{:confluence => 1, google_docs => 2}`. Then as we add more providers, we can refactor down the road as needed.

Comment (chewnoill) [reply lib/release_notes_bot/persists.ex:156]: Enums work, sure. 

Maybe more importantly though, if we only have one, we don't need to worry about it yet. Might as well drop the whole case statement, and just assume its confluence all the time.

Hardcoding on `provider_id` doesn't seem like the most stable way of managing this relationship though. This is a sequence field?

Comment (zach-bob) [reply lib/release_notes_bot/persists.ex:156]: You're right its not. My thoughts of this were that Confluence would always be the first (and right now only) entry in persistence_providers. Thats where I get provider_id from. Its Confluence's id in persistence_providers.

Comment (zach-bob) [reply lib/release_notes_bot/persists.ex:156]: Having this case statement has actually helped track down a few bugs where the relation between a project and its persistence_provider (it should be always confluence) was not there.

File: lib/release_notes_bot/persists.ex [hunk 2]
```diff
@@ -157,4 +220,12 @@ defmodule ReleaseNotesBot.Persists do
   defp replace_spaces_with_plus_signs(string), do: String.replace(string, " ", "+")
 
   defp drop_question_mark(string), do: String.replace(string, "?", "")
+
+  defp update_confluence_page() do
+    {:error, "Not implemented"}
+    # curl -u admin:admin -X PUT -H 'Content-Type: application/json' -d '{"id":"3604482","type":"page",
+    # "title":"new page","space":{"key":"TST"},"body":{"storage":{"value":
+    # "<p>This is the updated text for the new page</p>","representation":"storage"}},
+    # "version":{"number":2}}' http://localhost:8080/confluence/rest/api/content/3604482
```
Comment (sethkrasnianski) [inline lib/release_notes_bot/persists.ex]: If this is an artifact for testing code, could you identify it as such. Otherwise it could be mistaken for dead code.

File: lib/release_notes_bot/persists.ex [hunk 3]
```diff
@@ -68,18 +69,63 @@ defmodule ReleaseNotesBot.Persists do
...
+            {:ok, res}
+
+          {:error, err} ->
+            {:error, err}
+        end
+
+      _ ->
+        {:error, "Invalid action"}
+    end
+  end
```
Comment (sethkrasnianski) [inline lib/release_notes_bot/persists.ex:113]: We could alternatively write this as a pattern matched function instead of leveraging a simple case statement.

Each signature would capture the action name, for example 
```ex
choose_action(changes, "published", prov_id)`
``` 
and return the subsequently desired call. Otherwise we could fallback to a catchall like 
```ex
choose_action(_, _, _), do:   {:error, "Invalid action"}`
```
 From what Craig J told me, it takes advantage of compile time optimizations of the language.

File: lib/release_notes_bot/persists.ex [hunk 4]
```diff
@@ -221,7 +221,15 @@ defmodule ReleaseNotesBot.Persists do
...
-  defp update_confluence_page() do
+  defp get_headers(token) do
+    [
+      {"Content-Type", "application/json"},
+      {"Authorization", "Basic #{token}"}
+    ]
+  end
+
+  defp update_confluence_page(data) do
+    headers = get_headers(data.token)
```
Comment (sethkrasnianski) [inline lib/release_notes_bot/persists.ex]: This should be considered dead code

File: lib/release_notes_bot/persists.ex [hunk 5]
```diff
@@ -221,7 +221,15 @@ defmodule ReleaseNotesBot.Persists do
 
   defp drop_question_mark(string), do: String.replace(string, "?", "")
 
-  defp update_confluence_page() do
+  defp get_headers(token) do
+    [
+      {"Content-Type", "application/json"},
+      {"Authorization", "Basic #{token}"}
```
Comment (sethkrasnianski) [inline lib/release_notes_bot/persists.ex:302]: For functions like these, that setup data structures based off input, we should provide a data structure _as_ input so it's not only more flexible in the future but it's clear to the caller what we're passing.
Instead we could do
```ex
defp get_headers(%{"token" => token}) do
#
end

# more explicit caller
get_headers(%{"token" => data.token})

# future iterations don't have to worry about argument order
get_headers(%{"cache_control" => "no-cache", "token" => token})
```

File: lib/release_notes_bot/persists.ex [hunk 6]
```diff
@@ -193,6 +194,15 @@ defmodule ReleaseNotesBot.Persists do
...
 
+  defp update_confluence_page(data) do
+    headers = get_headers(data.token)
+    {:error, "Not implemented"}
+    # curl -u admin:admin -X PUT -H 'Content-Type: application/json' -d '{"id":"3604482","type":"page",
+    # "title":"new page","space":{"key":"TST"},"body":{"storage":{"value":
+    # "<p>This is the updated text for the new page</p>","representation":"storage"}},
+    # "version":{"number":2}}' http://localhost:8080/confluence/rest/api/content/3604482
+  end
+
```
Comment (sethkrasnianski) [inline lib/release_notes_bot/persists.ex:238]: Let's be conscious as to not randomly move code in unrelated commits. If this is truly the location you desire for this function, then introduce it here initially.

File: lib/release_notes_bot_web/controllers/webhook_controller.ex [hunk 7]
```diff
@@ -144,13 +170,18 @@ defmodule ReleaseNotesBotWeb.WebhookController do
...
+         page_info \\ nil
+       ) do
     # TO DO: Persist to all persistence providers
     case Persists.persist(
            build_persistence_title(release),
            release["body"],
            Enum.take(project_provider, 1) |> List.first(),
-           action
+           action,
+           page_info
```
Comment (sethkrasnianski) [inline lib/release_notes_bot_web/controllers/webhook_controller.ex:184]: Does this change belong in the commit where you introduce `page_info` to the `persist/5` function?

File: lib/release_notes_bot/persists.ex (hunk 1 above)
Comment (chewnoill) [reply lib/release_notes_bot/persists.ex:156]: I'm not against having a case statement, but having it be on the id field seems like a not great long term plan

File: lib/release_notes_bot/persists.ex (hunk 5 above)
Comment (zach-bob) [reply lib/release_notes_bot/persists.ex:302]: This is sexy - I'm definitely taking a page from this :)

File: lib/release_notes_bot/persists.ex (hunk 2 above)
Comment (zach-bob) [reply lib/release_notes_bot/persists.ex]: This will be purged - it was helpful during development but is no longer necessary

File: lib/release_notes_bot/release_tags.ex [hunk 8]
```diff
@@ -14,6 +14,26 @@ defmodule ReleaseNotesBot.ReleaseTags do
     |> Enum.member?(release_tag_id)
   end
 
+  @spec get_id(Integer.t(), String.t()) :: Interger.t()
+  def get_id(repo_id, release_tag_id) do
```
[the remark this answers is not in the document]
Comment (zach-bob) [reply lib/release_notes_bot/release_tags.ex:18]: Noted. Made this atomic in commit d4ade230bb1dbe2d31a30270e8dc5ad995e3e129

File: lib/release_notes_bot_web/controllers/webhook_controller.ex [hunk 9]
```diff
@@ -137,21 +142,18 @@ defmodule ReleaseNotesBotWeb.WebhookController do
   end
 
   defp determine_persistence(%{"release" => release, "action" => action}, project_provider) do
-    # Persist to all persistence providers
-    if action in @persist_actions do
```
[the remark this answers is not in the document]
Comment (zach-bob) [reply lib/release_notes_bot_web/controllers/webhook_controller.ex:141]: Yes - this guard was redundant :/ There were some plans to remove `@persist-actions` before but this change eliminates the rest of its uses.

File: lib/release_notes_bot_web/controllers/webhook_controller.ex (hunk 7 above)
Comment (zach-bob) [reply lib/release_notes_bot_web/controllers/webhook_controller.ex]: This change triggers the new `page_info` arg in the `persist/5` function. I can squash these together, but I felt that it was more readable to split these changes up when I originally putting this PR together.

File: lib/release_notes_bot/persists.ex (hunk 1 above)
Comment (zach-bob) [reply lib/release_notes_bot/persists.ex:156]: You've got it. I'll make a ticket for a refac

File: lib/release_notes_bot_web/controllers/webhook_controller.ex (hunk 7 above)
Comment (zach-bob) [reply lib/release_notes_bot_web/controllers/webhook_controller.ex]: For the sake of atomicness - it may make sense to just squash

File: lib/release_notes_bot/persists.ex (hunk 3 above)
Comment (sethkrasnianski) [reply lib/release_notes_bot/persists.ex]: Is this really resolved? It's not necessary for merging, I'd say. But if we don't want to make the change immediately, let's preserve this conversation as unresolved for posterity

Comment (zach-bob) [reply lib/release_notes_bot/persists.ex]: Yessir - I went into this more in the next PR. Its not yet optimal - but can definitely refac more

**Review by sethkrasnianski — APPROVED** [review]

**Review by chewnoill — APPROVED** [review]
🚀

**Comment by sethkrasnianski** [comment]
🪨 🇮🇹

**Comment by zach-bob** [comment]
🪨 
⛽
